### PR TITLE
Add ControlMsrE2

### PR DESCRIPTION
**Pull request by derBrumbaer**
Shall replace VerifyMsrE2 and allow changing the BIOS Option: CFGLock.

### CODE REVIEW

**Comment by vit9696** [comment]
Is that a WIP PR or does it need review? Currently it has a lot of codestyle issues, but if necessary, I could point them out. Thanks!

**Comment by derBrumbaer** [comment]
Thanks, that would be very helpful.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+#define CONTEXTS_MAX 8
+
+EFI_STATUS WalkListHeaders (
+                            IN EFI_HII_HANDLE * handles,
+                            IN EFI_HII_PACKAGE_LIST_HEADER **ListHeaders,
+                            IN UINT32 ListHeaderCount,
+                            IN EFI_STRING SearchString
+                            ) {
+    UINT16           OptionsCount;
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:24]: The indentation of the code should be made with **2 spaces**, and the function header should be formatted as follows:

```C
EFI_STATUS
WalkListHeaders (
  IN EFI_HII_HANDLE                *Handles,
  IN EFI_HII_PACKAGE_LIST_HEADER   **ListHeaders,
  IN UINT32                        ListHeaderCount,
  IN EFI_STRING                    SearchString
  )
{
  /* … */
}
```

Note the **newlines** as well as argument **indentation** and **alignment**. Also braces and curly-braces positioning.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    UINT16           ContextsCount;
+    OneOfContext     Contexts[CONTEXTS_MAX];
+    UINT32           ListHeaderIndex;
+    
+    EFI_STATUS Status = EFI_SUCCESS;
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:29]: C89-style stack variable declaration is used with the only used scope being top function scope. So every stack variable is put **to the beginning** of the function and is **not** initialised there.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    
+    OptionsCount = 0;
+    ContextsCount = 0;
+    ListHeaderIndex = 0;
+    
+    // For Each Handle
+
+    for (EFI_HII_HANDLE* h = (EFI_HII_HANDLE*) handles; (*h != NULL) && (ContextsCount < CONTEXTS_MAX); h++, ListHeaderIndex++) {
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:37]: Similarly to other variable declaration restrictions, C99-like loop variable declaration is now allowed.

Also, every stack variable should be **camel-cased** and starting with a **capital** letter, e.g. `Handle`.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+        ListHeaders[ListHeaderIndex] = HiiExportPackageLists(*h);
+        
+        if (ListHeaders[ListHeaderIndex] != NULL) {
+            
+            if (IS_VERBOSE()) {
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:43]: For both functions and macros the argument braces should be padded with spaces, like `IS_VERBOSE ()` or `HiiExportPackageLists (*Handle)`.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            while (ContextsCount < CONTEXTS_MAX) {
+
+                if (IS_VERBOSE()) {
+                    Print (L"Package Type: %02X ", PkgHeader->Type);
+                }
+                
+                if (PkgHeader->Type == EFI_HII_PACKAGE_END) {
+                    break;
+                }
+                else if (PkgHeader->Type == EFI_HII_PACKAGE_FORMS) {
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:62]: For ifs the no-newline style is used:

```C
if (condition) {
  /* … */
} else if (condition) {
  /* … */
} else {
  /* … */
}
```

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                                Contexts[ContextsCount].efiHandle = *h;
+                                Contexts[ContextsCount].listHeader = ListHeaders[ListHeaderIndex];
+                                Contexts[ContextsCount].pkgHeader = PkgHeader;
+                                Contexts[ContextsCount].firstIfrHeader = PADD(IfrHeader, IfrHeader->Length);
+                                Contexts[ContextsCount].ifrVarStore = NULL;
+                                Contexts[ContextsCount].ifrOneOf = NULL;
+                                Contexts[ContextsCount].stopAt = DONT_STOP_AT;
+                                Contexts[ContextsCount].count = OptionsCount;
+                                
+                                DoForEachOpCode (Contexts[ContextsCount].firstIfrHeader, EFI_IFR_ONE_OF_OP, NULL, &Contexts[ContextsCount], HandleOneOf);
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:94]: When the amount of arguments is too long, it is preferable to split them one argument per line, like this with optional explanation if necessary:

```C
DoForEachOpCode (
  Contexts[ContextsCount].firstIfrHeader,
  EFI_IFR_ONE_OF_OP,
  NULL,                               ///< NULL blahblahblah
  &Contexts[ContextsCount],
  HandleOneOf
  );
```

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        
+        if (ListHeaders[ListHeaderIndex] != NULL) {
+            
+            if (IS_VERBOSE()) {
+                Print (L"Package List: ");
+                PrintGuid (&ListHeaders[ListHeaderIndex]->PackageListGuid);
+                Print (L"\n");
+            }
+            
+            // First package in list
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:49]: Comments are put into blocks with three-lines or are placement comments.

```C
//
// First package in the list
//
```

or 

```C
}  ///< For Each Handle
```

File: Application/ControlMsrE2/ControlMsrE2.h [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    EFI_HII_HANDLE               efiHandle;     // Handle of the list the Form is part of
+    EFI_HII_PACKAGE_LIST_HEADER *listHeader;    // Ptr to the contents of the handle
+    EFI_HII_PACKAGE_HEADER      *pkgHeader;     // Ptr to the current package in the list
+    EFI_IFR_OP_HEADER           *firstIfrHeader; // Ptr to first IfrHeader in Package
+    EFI_IFR_VARSTORE            *ifrVarStore;   // Ptr to ifrHeader of corresponding varstore in same list
+    EFI_IFR_ONE_OF              *ifrOneOf;      // Ptr to ifrHeader of BIOS Option
+    UINT16                       stopAt;        // Used to find a certain BIOS Option
+    UINT16                       count;         // Running number of suitable BIOS Options
+    
+} OneOfContext;
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.h:65]: Structure types are named with capital snake case. E.g. `ONE_OF_CONTEXT`.

File: Application/ControlMsrE2/ControlMsrE2.h [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+/*
+ Callback - What to do, when a ifrHeader with a defined OP_CODE is found
+ ifrHeader   Current ifrHeader under scrutiny
+ stop        Ptr to Stop flag. If true no further search for opcodes. Can be NULL.
+ context     Ptr to Hanlder specific data
+ */
+typedef VOID OpCodeHandler (EFI_IFR_OP_HEADER* ifrHeader, UINT8* stop, VOID* context);
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.h:73]: Same for function types, which should also abide to general function declaration rules:

```C
/**
  Callback - What to do, when a IfrHeader with a defined OP_CODE is found.

  @param[in] IfrHeader   Current ifrHeader under scrutiny.
  @param[in] Stop        Pointer to Stop flag. If TRUE no further search for opcodes, optional.
  @param[in] Context     Ptr to Hanlder specific data.
**/
typedef
VOID
OP_CODE_HANDLER (
  IN EFI_IFR_OP_HEADER  *IfrHeader,
  IN UINT8              *Stop  OPTIONAL,
  IN VOID               *Context
  );
```

File: Application/ControlMsrE2/ControlMsrE2.h [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    ARG_LOCK = 1,
+    ARG_UNLOCK = 2,
+    ARG_CHECK = 4,
+    ARG_INTERACTIVE = 8,
+    ARG_VERBOSE = 16
+};
+
+#define PADD(x,y) (void*)(((char*) x) + y)
+
+#define IS_VERBOSE() (mFlags & ARG_VERBOSE)
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.h:88]: For all non-boolean operations it is recommended to explicitly mark the operation result. I.e.  `((mFlags & ARG_VERBOSE) != 0)`.

File: Application/ControlMsrE2/ControlMsrE2.h [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+
+// Wait for Keypress of Y or N. Ignores case.
+extern UINT32 ReadYN ();
+// Wait for any key press
+extern CHAR16 ReadAnyKey ();
+
+// Parse commandline arguments
+extern UINTN InterpretArguments ();
+extern VOID  PrintUINT8Str (IN UINT8 n[1]);
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.h:111]: Please use `IN UINT8 *String`.

File: Application/ControlMsrE2/ControlMsrE2.c [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+            EFI_HII_PACKAGE_HEADER* PkgHeader = PADD(ListHeaders[ListHeaderIndex], sizeof(EFI_HII_PACKAGE_LIST_HEADER));
+
+            // For each package in list
+            while (ContextsCount < CONTEXTS_MAX) {
+
+                if (IS_VERBOSE()) {
+                    Print (L"Package Type: %02X ", PkgHeader->Type);
```
Comment (vit9696) [inline Application/ControlMsrE2/ControlMsrE2.c:56]: When dealing with user output it is often more useful to perform printing via `DebugLib`, as it allows to not only present things to user but also save the transcript to the logfile. For user interaction `DEBUG_WARN` level is recommended. I.e. something like this:

```C
  DEBUG ((DEBUG_WARN, "Package Type: %02X ", PkgHeader->Type));
```

The nuance here is that only one-liners are allowed, so if you think this does not fit you, feel free to keep using `Prints` as before.

**Comment by derBrumbaer** [comment]
I already spent too much time on getting the tool compiling on compilers I will never use. 
I will not spent any more time on transferring it to the OpenCore project.

I understand what your goal is and why you need to do what you. 
But we do not have the same goal and I'm not striving to be part of acidanthera.

The tool has a primitive algorithm and you will write the tool anew in less time than I need to getting it up to your standards and you reviewing it again and again.

I apologise for wasting your time.

**Comment by 1alessandro1** [comment]
All this work gone only for some codestyle issues? @derBrumbaer @vit9696 
Come on, many people can benefit from this tool which may become very useful for many and avoids lots of headaches especially on laptops where very often finding the offset with IFRExtract and UEFITool and applying it with setup_var doesn't work, please take your time to fix those one by one even in a month or so, just a little effort twice a week or so...

**Comment by vit9696** [comment]
@1alessandro1 it is not gone as the changes are not lost. We indeed cannot merge the code changes that do not adhere to OpenCore coding standards. Sooner or later somebody should address this hopefully. Having the draft implementation is rather useful.

**Comment by SukkaW** [comment]
@vit9696 Well, should an issue being opened at the bugtracker so that somebody might trying to implement it again?

**Comment by vit9696** [comment]
As a help-wanted, I guess.

**Pull request closed without merging**